### ISSUE TITLE
NULL now has a sub-type(?), breaking NULL comparisons with it

### ISSUE THREAD

**Issue by Kefta** [Facepunch/garrysmod-issues#3089]
Consider the following code:
```
-- In a weapon
Player.LastWeapon = self
self:Remove()

-- In a hook
print(Player.LastWeapon == Player:GetActiveWeapon())
```

This now prints false since LastWeapon is a different NULL-type (tostring returns Weapon [NULL]) while GetActiveWeapon returns NULL. This can be fixed by making GetActiveWeapon return the Weapon NULL type, or making all NULL handles the same type.

Excuse the rough terminology; don't really know how Lua's NULL is defined in the engine.

**Comment by willox** [comment]
doesn't this happen in the main branch as well?

**Comment by Kefta** [comment]
This does happen on the main branch, as well.

**Comment by willox** [comment]
i'm not sure comparisons to NULL are going to ever be reliable without changes to LuaJIT or wrecking compatibility with a lot of addons (i don't want to do that)

**Comment by NeatNit** [comment]
>we can set the userdatas' metatable to Entity when they become invalid perhaps

I don't think that's a good idea, when you use nullwep:SomeWeaponFunction() it would throw an 'attempt to call nil value' instead of 'attempt to use null entity/weapon'. I think. Or does the latter error get thrown when indexing NULL entities?

**Comment by Kefta** [comment]
@willox I think that would be suitable. Since NULL is just supposed to act as an invalid entity pointer, it shouldn't have a specific userdata type that would differentiate one NULL from another
@NeatNit I agree that the errors would somewhat obscure the issue, but I believe it would still be clear that validity checks need to be added

**Comment by NeatNit** [comment]
If we do that we might as well make null throw an error on index regardless of the key, just like a regular null pointer is supposed to do

**Comment by Kefta** [comment]
I think that the benefit of having it be userdata in Lua is being able to let methods like IsValid, IsNPC, etc. not error and instead just return false, making logic much simpler.

**Comment by NeatNit** [comment]
Yes, but besides IsValid, any other index should error. I mean, a null entity (or null physobj, etc) is one that has either been removed or never existed in the first place. Even basic data storing like `ent.SomeValue = 3 print(ent.SomeValue)` should throw an error.

**Comment by NeatNit** [comment]
Actually, I have no idea how null comparisons should work which is the main topic of this issue. The thing I brought up is irrelevant.

Should null comparisons ever return true? In SQL null comparisons are ALWAYS null, not even true or false, which is a little shit but it does force you to think differently​ about NULL. the only way you can check for null is IS NULL or IS NOT NULL. In gmod I suppose the only correct way to check for Null is IsValid(val). Null-to-null comparison is ill-defined IMO.

**Comment by Kefta** [comment]
Well in C/C++, NULL (which is defined as just 0 in many compilers)/nullptr == NULL/nullptr. Since that's where glua obviously derives its NULL implementation, I don't see any reason NULL != NULL.

**Comment by Kefta** [comment]
Actually, changing NULL's __index to error if the key isn't IsValid would be a beneficial change to let people know to add validity checks.

**Comment by Kefta** [comment]
Can you not just change Entity.__eq to return true if both entities are NULL?

**Comment by thegrb93** [comment]
Remove() on an entity doesn't change it to NULL. I think the distinction is important for some addons.

**Comment by Kefta** [comment]
It's considered invalid when it's removed on the next tick which is why __tostring returns ``EntityType [NULL]``. It should be expected behaviour for ``ent == NULL`` when ``not ent:IsValid() and not ent == game.GetWorld()``. Thus __eq should:
- Return true if both entities are invalid (actually invalid using the internal bool, not IsValid since that would fail for the world)
- Return false if either entity is invalid but the other isn't
- Otherwise return if entptr1 == entptr2
This kind of complex behaviour is due to how the NULL-object paradigm is implemented here, but the entity userdata object is essentially acting like a handle instead of a pointer. When the entity is removed, it disappears and essentially becomes NULL. What addons/code would break from this change?

**Comment by thegrb93** [comment]
My point is there are addons that expect NULL to be world and nothing else. Once you start throwing removed entities into the mix, bad stuff can happen. For example, I remember one addon had an exploit where an invalid player was determined to be the console and have superadmin access, so to exploit it, the player could leave the game and, with some kind of entity they left behind, invoke a superadmin command. Addons that use NULL for this purpose would suffer the same consequences.

**Comment by Kefta** [comment]
> My point is there are addons that expect NULL to be world and nothing else.

That's not how it behaves now though, removing an entity that has the entity metatable then comparing it to NULL works properly. This only applies to entities with non-Entity metatables like players, weapons, and NPCs.

> For example, I remember one addon had an exploit where an invalid player was determined to be the console and have superadmin access, so to exploit it, the player could leave the game and, with some kind of entity they left behind, invoke a superadmin command.

But NULL is the console - the concommand callback will never give you an invalid entity unless it's being invoked from the console. If this is indeed how the exploit worked, it sounds like they were storing the player entity til the next tick when it would go invalid before checking if they were superadmin. Which doesn't even make sense since invalid players have never equaled NULL, so if this exploit was real, it was the script's fault anyway. I really don't see what code this would break considering this behaviour only applies to Lua subtypes of entities, which makes it appear more like a design oversight than intended behaviour.

**Comment by thegrb93** [comment]
the exploit was because IsValid was used, not NULL. yeah it was the scripts fault, but if ==NULL starts working for disconnected players then it might return.

**Comment by Kefta** [comment]
Direct ``NULL`` comparisons are hardly used now compared to the global ``IsValid``/member ``IsValid`` (I don't think they're used at all in the base code), and I can't think of any cases where a ``NULL`` comparison would even be used on a formally valid player anyway since it never functioned. Perhaps there would be some very suspect use cases that would start to behave differently, but they were relying on undefined behaviour anyway.

**Comment by thegrb93** [comment]
Starfall currently uses it, but I'll probably be changing it just in case.

**Comment by Kefta** [comment]
Could you link the use case?

**Comment by thegrb93** [comment]
https://github.com/thegrb93/StarfallEx/blob/master/lua/starfall/instance.lua#L61

**Comment by thegrb93** [comment]
https://github.com/thegrb93/StarfallEx/blob/master/lua/starfall/libs_sv/prop.lua#L138

**Comment by Kefta** [comment]
Yep, exactly as expected:
```Lua
] lua_run e = Entity(1):GetEyeTrace().Entity
> e = Entity(1):GetEyeTrace().Entity...
] lua_run e:Remove()
> e:Remove()...
] lua_run print(e)
> print(e)...
[NULL NPC]
] lua_run print(e == NULL, FindMetaTable("Entity").__eq(e, NULL), NULL == e, FindMetaTable("Entity").__eq(NULL, e))
> print(e == NULL, FindMetaTable("Entity").__eq(e, NULL), NULL == e, FindMetaTable("Entity").__eq(NULL, e))...
false	true	false	true
```

This is a bug with Entity subtypes' metatables. You can use the Entity's metatable __eq directly for correct comparrisons.

**Comment by Astralcircle** [comment]
> Yep, exactly as expected:
> 
> ] lua_run e = Entity(1):GetEyeTrace().Entity
> > e = Entity(1):GetEyeTrace().Entity...
> ] lua_run e:Remove()
> > e:Remove()...
> ] lua_run print(e)
> > print(e)...
> [NULL NPC]
> ] lua_run print(e == NULL, FindMetaTable("Entity").__eq(e, NULL), NULL == e, FindMetaTable("Entity").__eq(NULL, e))
> > print(e == NULL, FindMetaTable("Entity").__eq(e, NULL), NULL == e, FindMetaTable("Entity").__eq(NULL, e))...
> false	true	false	true
> This is a bug with Entity subtypes' metatables. You can use the Entity's metatable __eq directly for correct comparrisons.

Is there any reason why this can't be done in game? Seems to be pretty simple
